File: app/src/main/java/com/stevenfrew/beatprompter/bluetooth/ChooseSongMessage.java
package com.stevenfrew.beatprompter.bluetooth;

import android.util.Log;

import com.stevenfrew.beatprompter.BeatPrompterApplication;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

public class ChooseSongMessage extends BluetoothMessage
{
    static final byte CHOOSE_SONG_MESSAGE_ID=0;

    public String mTitle;
    public String mTrack;
    public boolean mBeatScroll;
    public boolean mSmoothScroll;
    public int mOrientation;
    public int mMinFontSize;
    public int mMaxFontSize;
    public int mScreenWidth;
    public int mScreenHeight;

    public ChooseSongMessage(String title,String track,int orientation,boolean beatScroll,boolean smoothScroll,int minFontSize,int maxFontSize,int screenWidth,int screenHeight)
    {
        mTitle=title==null?"":title;
        mTrack=track==null?"":track;
        mOrientation=orientation;
        mBeatScroll=beatScroll;
        mSmoothScroll=smoothScroll;
        mMinFontSize=minFontSize;
        mMaxFontSize=maxFontSize;
        mScreenWidth=screenWidth;
        mScreenHeight=screenHeight;
    }

    ChooseSongMessage(byte[] bytes) throws NotEnoughBluetoothDataException
    {
        try
        {
            final ByteArrayInputStream byteArrayInputStream =
                    new ByteArrayInputStream(bytes);
            int dataRead=byteArrayInputStream.read(new byte[1]);
            if(dataRead==1) {
                int availableStart=byteArrayInputStream.available();
                final ObjectInputStream objectInputStream =
                        new ObjectInputStream(byteArrayInputStream);
                mTitle = (String) objectInputStream.readObject();
                mTrack = (String) objectInputStream.readObject();
                mBeatScroll = objectInputStream.readBoolean();
                mSmoothScroll = objectInputStream.readBoolean();
                mOrientation = objectInputStream.readInt();
                mMinFontSize = objectInputStream.readInt();
                mMaxFontSize = objectInputStream.readInt();
                mScreenWidth=objectInputStream.readInt();
                mScreenHeight=objectInputStream.readInt();
                int availableEnd=byteArrayInputStream.available();
                objectInputStream.close();
                byteArrayInputStream.close();
                mMessageLength=1+(availableStart-availableEnd);
            }
            else
                throw new NotEnoughBluetoothDataException();
        }
        catch(Exception e)
        {
            Log.e(BeatPrompterApplication.TAG,"Couldn't read ChooseSongMessage data, assuming not enough data",e);
            throw new NotEnoughBluetoothDataException();
        }
    }

    public byte[] getBytes()
    {
        try
        {
            final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            byteArrayOutputStream.write(new byte[]{CHOOSE_SONG_MESSAGE_ID},0,1);
            final ObjectOutputStream objectOutputStream =
                    new ObjectOutputStream(byteArrayOutputStream);
            objectOutputStream.writeObject(mTitle);
            objectOutputStream.writeObject(mTrack);
            objectOutputStream.writeBoolean(mBeatScroll);
            objectOutputStream.writeBoolean(mSmoothScroll);
            objectOutputStream.writeInt(mOrientation);
            objectOutputStream.writeInt(mMinFontSize);
            objectOutputStream.writeInt(mMaxFontSize);
            objectOutputStream.writeInt(mScreenWidth);
            objectOutputStream.writeInt(mScreenHeight);
            objectOutputStream.flush();
            objectOutputStream.close();

            final byte[] byteArray = byteArrayOutputStream.toByteArray();
            byteArrayOutputStream.close();

            return byteArray;
        }
        catch(Exception e)
        {
            Log.e(BeatPrompterApplication.TAG,"Couldn't write ChooseSongMessage data.",e);
        }
        return null;
    }
}

File: app/src/main/kotlin/com/stevenfrew/beatprompter/bluetooth/ChooseSongMessage.kt
package com.stevenfrew.beatprompter.bluetooth

import android.graphics.Rect
import android.util.Log
import com.stevenfrew.beatprompter.BeatPrompterApplication
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

/**
 * Message that is sent/received when a song is chosen.
 */
class ChooseSongMessage constructor(val mNormalizedTitle: String, val mNormalizedArtist:String, val mTrack: String, val mOrientation: Int, val mBeatScroll: Boolean, val mSmoothScroll: Boolean, val mMinFontSize: Float, val mMaxFontSize: Float, val mScreenSize: Rect): BluetoothMessage() {
    override val bytes: ByteArray
        get() {
            return ByteArrayOutputStream().apply {
                write(byteArrayOf(CHOOSE_SONG_MESSAGE_ID), 0, 1)
                ObjectOutputStream(this).apply {
                    writeObject(mNormalizedTitle)
                    writeObject(mNormalizedArtist)
                    writeObject(mTrack)
                    writeBoolean(mBeatScroll)
                    writeBoolean(mSmoothScroll)
                    writeInt(mOrientation)
                    writeFloat(mMinFontSize)
                    writeFloat(mMaxFontSize)
                    writeInt(mScreenSize.width())
                    writeInt(mScreenSize.height())
                    flush()
                    close()
                }
            }.toByteArray()
        }

    companion object {
        internal const val CHOOSE_SONG_MESSAGE_ID: Byte = 0

        @Throws(NotEnoughBluetoothDataException::class)
        internal fun fromBytes(bytes: ByteArray):IncomingBluetoothMessage {
            try {
                ByteArrayInputStream(bytes).apply {
                    val dataRead = read(ByteArray(1))
                    if (dataRead == 1) {
                        val availableStart = available()
                        ObjectInputStream(this).apply {
                            val title = readObject() as String
                            val artist = readObject() as String
                            val track = readObject() as String
                            val beatScroll = readBoolean()
                            val smoothScroll = readBoolean()
                            val orientation = readInt()
                            val minFontSize = readFloat()
                            val maxFontSize = readFloat()
                            val screenWidth = readInt()
                            val screenHeight = readInt()
                            val availableEnd = available()
                            val messageLength = 1 + (availableStart - availableEnd)
                            close()
                            return IncomingBluetoothMessage(ChooseSongMessage(title,artist,track,orientation,beatScroll,smoothScroll,minFontSize,maxFontSize,Rect(0,0,screenWidth,screenHeight)),messageLength)
                        }
                    }
                }
            } catch (e: Exception) {
                Log.e(BeatPrompterApplication.TAG, "Couldn't read ChooseSongMessage data, assuming not enough data", e)
            }
            throw NotEnoughBluetoothDataException()
        }
    }
}
